option to restore previous TAB behavior

tpapp opened the pull request:
Fixes #193.

This is preliminary, **comments are welcome**, even along the lines of "Tamás your elisp code is so horrible that I had to wash my eyes with soap after reading it, here is how you do it nicer".

Try it with
```emacs-lisp
(define-key julia-mode-map (kbd "TAB") 'julia-latexsub-or-indent)
```
Maybe we can add an option for it.

Possible improvements:

- [ ] use something else than `completing-read`? is there a nicer popup version?
- [ ] annotate with the actual replacement in `completing-read`
- [x] option for greedy replacement for a full match (eg `\bar` will replace, no popup)
- [ ] option to pick the first replacement, whatever it is

Have to do before merging this:

- [x] some unit tests
- [ ] code review from someone
- [x] some testing by interested users

torfjelde commented:
I can be one of those "interested users":) Currently trying it out, and it seems to be working nicely :) Thanks! @tpapp !

dhanak commented on julia-mode.el:
```diff
@@ -908,6 +919,40 @@ buffer where the LaTeX symbol starts."
+    (let ((partial (buffer-substring-no-properties beg (point))))
+      (message "partial %s" partial)
+      (when-let (replacements (gethash partial julia-mode--latexsubs-partials))
+        (message "replacements %s" (prin1 replacements))
```
This is cool for debugging purposes, but it bloats the minibuffer unnecessarily, even if temporarily.

This and the previous message line should probably be deleted before merging.

dhanak commented on julia-mode.el:
```diff
@@ -908,6 +919,40 @@ buffer where the LaTeX symbol starts."
+        (message "replacements %s" (prin1 replacements))
+        (let* ((complete-match (member partial replacements))
+               (replacement (cond ((and complete-match julia-latexsub-greedy) partial)
+                                  ((cdr replacements) (gethash (completing-read "LaTeX completions: " replacements) julia-mode-latexsubs))
```
I think this is an error.

When the completion is not full, e.g., \cir<TAB>, I am offered a list of valid completions. When I press enter on one of them, I get a `progn: Wrong type argument: char-or-string-p, nil` error.

After some inspection, it seems that this is because this line here should return the raw latex control sequence, not the already substituted value. The substitution is performed in the  `julia-latexsub-or-indent` function.

If I change this line to

```elisp
((cdr replacements) (completing-read "LaTeX completions: " replacements))
```

then it works as expected.

dhanak reviewed (changes requested):
I tried this briefly and apart from the two comments below, it seems to work as expected. Nice work, and thanks for the effort!

dhanak commented:
Any chance of getting this merged soon?

tpapp commented:
It still needs tests. I will try to do them soon and merge your suggestions too. I appreciate the ping.

tpapp commented:
I incorporated the suggestions by @dhanak and added some basic tests. While one can always prettify this further, it is functional so I would prefer to merge and then leave bells & whistles for future work.

Further testing is welcome, but a code review would be best.

tpapp commented:
asking the usual suspects for review before merging, @giordano, @non-Jedi, @FelipeLema, @ajhalme

non-Jedi commented on julia-mode.el:
```diff
@@ -919,6 +941,42 @@ buffer where the LaTeX symbol starts."
+
+eg in your `julia-mode-hook'."
+  (interactive "*i")
+  (if-let (replacement (julia-mode--latexsub-before-point))
```
I guess this option is missing this from the OP

> option for greedy replacement for a full match (eg \bar will replace, no popup)

but I think you could handle just that case with significantly less complexity than this PR.

tpapp replied:
Sorry, I do not understand. `julia-latexsub-greedy`, which is set to `t` by default, does exactly this (eg `\bar` just completes).

Regarding complexity: I agree that in an ideal world, this PR would not have been necessary. But practically, the generic completion mechanisms of Emacs have proven to be baroque and full of corner cases and/or implementation bugs like #204 (thanks for chasing many of those down!). I think that for now there is value in maintaining our own version which Just Works :tm:.

non-Jedi commented on julia-mode.el:
```diff
@@ -919,6 +941,42 @@ buffer where the LaTeX symbol starts."
+      (progn
+        (delete-backward-char (- (point) (car replacement)))
+        (insert (gethash (cdr replacement) julia-mode-latexsubs)))
+    (julia-indent-line)))
```
Wait, if this ultimately calls `julia-indent-line`, I don't think it solves the problem from #193 at all. The issue in #193 was being able to have `TAB` work for both latex substitution as well as inserting `TAB`s (see example in this comment). I think this might do what ronisbr was looking for if it called `indent-for-tab-command` instead of `julia-indent-line`.

tpapp replied:
Sorry, I missed this comment, just found it among messages. This PR solves the problem for me, if anyone still has problems with it please open another issue.

Specifically, it only calls `julia-indent-line` when it could not make a substitution. So, TAB tries to subtitute, and indents if it can't.  I find that behavior useful.

non-Jedi replied:
So basically it gives you the reverse of the behavior of `indent-for-tab-command` when `tab-always-indent` is set to `'complete`:

- With `indent-for-tab-command` with `(setq tab-always-indent 'complete)`: indent then complete
- With this function: complete (but only latexsubs) then indent

If we change `julia-indent-line` in this function to `indent-for-tab-command`, the user can change between TAB inserting tabs and TAB indenting the line by customizing the value of `tab-always-indent`. I'll create a PR for your review when I have a few minutes.

If you haven't seen it yet, could you please comment on [my other review comment when you get a chance](https://github.com/JuliaEditorSupport/julia-emacs/pull/196#discussion_r1592636480)?
